### ISSUE TITLE
Exception caused by Handle.GET crashes the database

### ISSUE THREAD

**Issue by warpech** [Starcounter/Starcounter#3128]
A below exception, which is likely to occur during development, cases a critical error and crashes the database:

![image](https://cloud.githubusercontent.com/assets/566463/11525385/fa3a0bea-98d2-11e5-8dd3-f7e56b795204.png)

From error log:

```
12/2/2015 9:00:00 AM	Critical	#sc://desktop-itup0ci/personal/default	Starcounter	"Starcounter.DbException: ScErrFailingEntrypoint (SCERR4259): The user code entrypoint method raised an exception. System.ArgumentException: Trying to add a delegate to an already existing handler: GET /braintree/app-icon  on port 8080
   at Starcounter.Rest.UserHandlerInfo.TryAddProxyOrReplaceDelegate(Func`4 userDelegate, HandlerOptions ho) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 381
   at Starcounter.Rest.UriHandlersManager.RegisterUriHandler(UInt16 port, String originalUriInfo, String processedUriInfo, Byte[] nativeParamTypes, Type messageType, Func`4 wrappedDelegate, NetworkProtocolType protoType, HandlerOptions ho) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 616
   at Starcounter.Rest.UriManagedHandlersCodegen.RegisterDelegate(UInt16 port, String methodSpaceUri, MethodInfo userDelegateInfo, Expression delegExpr, NetworkProtocolType protoType, HandlerOptions ho) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 619
   at Starcounter.Rest.UriManagedHandlersCodegen.GenerateParsingDelegate(UInt16 port, String methodSpaceUri, Func`1 userDelegate, HandlerOptions ho, NetworkProtocolType protoType) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 844
   at Starcounter.Rest.RestRegistrationProxy.RegisterHandler(UInt16 port, String methodSpaceUri, Func`1 handler, HandlerOptions ho) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Rest\RestInitialization.cs:line 59
   at Starcounter.Handle.GET(String uriTemplate, Func`1 code, HandlerOptions ho) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Internal\Handle.GET.cs:line 115
   at PaymentHistory.Handlers.LauncherHandlers.Register() in C:\www\handsoncode-apps\PaymentHistory\PaymentHistory\Handlers\LauncherHandlers.cs:line 13
   at PaymentHistory.Program.Main() in C:\www\handsoncode-apps\PaymentHistory\PaymentHistory\Program.cs:line 12
Version: 2.1.689.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4259.
   at Starcounter.Hosting.Package.ExecuteEntryPoint(Application application) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 580
   at Starcounter.Hosting.Package.ProcessWithinCurrentApplication(Application application) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 271
   at Starcounter.Hosting.Package.Process() in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 143
   at Starcounter.Hosting.Package.Process(IntPtr hPackage) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 56
   at StarcounterInternal.Hosting.Processor.RunMessageLoop(Void* hsched) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Hosting\Starcounter.Hosting\Processor.cs:line 71
ErrorCode=4259
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR4259
---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Starcounter.Hosting.Package.ExecuteEntryPoint(Application application) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Hosting\Starcounter.Hosting\Package.cs:line 554
HResult=-2146232828
---> System.ArgumentException: Trying to add a delegate to an already existing handler: GET /braintree/app-icon  on port 8080
   at Starcounter.Rest.UserHandlerInfo.TryAddProxyOrReplaceDelegate(Func`4 userDelegate, HandlerOptions ho) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 381
   at Starcounter.Rest.UriHandlersManager.RegisterUriHandler(UInt16 port, String originalUriInfo, String processedUriInfo, Byte[] nativeParamTypes, Type messageType, Func`4 wrappedDelegate, NetworkProtocolType protoType, HandlerOptions ho) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 616
   at Starcounter.Rest.UriManagedHandlersCodegen.RegisterDelegate(UInt16 port, String methodSpaceUri, MethodInfo userDelegateInfo, Expression delegExpr, NetworkProtocolType protoType, HandlerOptions ho) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 619
   at Starcounter.Rest.UriManagedHandlersCodegen.GenerateParsingDelegate(UInt16 port, String methodSpaceUri, Func`1 userDelegate, HandlerOptions ho, NetworkProtocolType protoType) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 844
   at Starcounter.Rest.RestRegistrationProxy.RegisterHandler(UInt16 port, String methodSpaceUri, Func`1 handler, HandlerOptions ho) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Rest\RestInitialization.cs:line 59
   at Starcounter.Handle.GET(String uriTemplate, Func`1 code, HandlerOptions ho) in C:\TeamCity\TeamCity8\buildAgent\work\sc-develop-nightly-109\Level1\src\Starcounter.Internal\Handle.GET.cs:line 115
   at PaymentHistory.Handlers.LauncherHandlers.Register() in C:\www\handsoncode-apps\PaymentHistory\PaymentHistory\Handlers\LauncherHandlers.cs:line 13
   at PaymentHistory.Program.Main() in C:\www\handsoncode-apps\PaymentHistory\PaymentHistory\Program.cs:line 12
HResult=-2147024809
"
```

**Comment by alemoi** [comment]
It says "GET /braintree/app-icon  on port 8080" is already registered. Do you have it in some other app? And it crashes the codehost I guess, not the database process.

**Comment by warpech** [comment]
Thanks for clarification. All I see as a user is that the database gets stopped in the Administrator.

**Comment by alemoi** [comment]
> Thanks for clarification. All I see as a user is that the database gets stopped in the Administrator.

But what about the duplicate handler? :) How it ended up there?

**Comment by warpech** [comment]
The developer copy pasted this by mistake. 

I only think that there is a flaw in Starcounter that such a simple problem causes the database to stop.

**Comment by alemoi** [comment]
> I only think that there is a flaw in Starcounter that such a simple problem causes the database to stop. It needs to be mistakeproof because people make mistakes.

You mean the codehost again :)

Anyways, in this case, when two same handlers are added, what do you expect from the codehost? Should it start? :)

**Comment by per-samuelsson** [comment]
> Anyways, in this case, when two same handlers are added, what do you expect from the codehost?

I've said this before: what I would expect as a developer is very simple: an exception saying I can't register such handler, becuase one with the same name is already registered.

```C#
try {
  Handle.GET("/foo") { () => {] }
}
catch (DublicateHandlerException e) {
  // Oh, that handler is already occupied. I can tell the user the service it provide is not
  // available.
}
```

Very similar in semantics to what happens when I add a key to a directory that already exist.

(Yes, this exact same case, we've seen before. Thats why I have a comment on it). :smile:

**Comment by per-samuelsson** [comment]
About that above comment: I think the implemented pattern is already pretty much in place. To make it perfect, we should map it to an exception I can distinguish, like above. But that's just a minor thing really. Simple to do (with the excpetion factory) and not really a bug if its not done, the more general argument exception is fine too.

**Comment by per-samuelsson** [comment]
@warpech 

Just to clarify our take here: what we do when there is a fault detected in an application **that is not handled** is to "throw the application out". 

Its very much in line with how a normal C# desktop application work. If an error occurs, and the application doesn't handle it (catch it, in exception-based lingo), it "bubbles up" out of the scope of the application, and the OS will terminate the process.

In Starcounter, since we are running multiple apps in the same database, we can do it a little bit differently - we can try keep the code host process stay alive (or rather, we recycle it: we terminate it, and launch a new one). If there are other apps running inside it, they should be restarted in that new host. (You are refering to this as "the database", and I think righly so, becuase thats what our users do too. But is actually just the process that contains the user code, as @alemoi points out).

How the web administrator handle this on the UI-level, I'm not sure. In what way are you starting the failing application? And are other applications already running?

**Comment by alemoi** [comment]
> About that above comment: I think the implemented pattern is already pretty much in place. To make it perfect, we should map it to an exception I can distinguish, like above.

Just curious about the existing distinguished exception types that we already throwing right now. I guess its not so many of them right? Probably covering quite a small subset of problem cases.

**Comment by alemoi** [comment]
I found the problem with current solution. I need to use ErrorCode.ToException. Fixing this now!

**Comment by per-samuelsson** [comment]
> I need to use `ErrorCode.ToException`.

Yeah, sorrry, I thought you new that. Thats the way we broadcast exceptions via the installed `ExceptionFactory`.

> I guess its not so many of them right? Probably covering quite a small subset of problem cases.

Right. And it shouldn't be that many either. There are good guides on exception best practices out there. One simple think I use to think about is how meaningful it is for a caller to be able to catch (and handle) a certain errornous case. If you can think of a good standard way to handle a particular error, then it might deserve an own exception type.

This case is borderline I would say. But having a dedicated one could allow clients to do something like I psedo-coded above I guess.

**Comment by alemoi** [comment]
Throwing error code SCERRHANDLERALREADYREGISTERED fixed via 747136d16e04f8b04bfd24f9aacf81af1e40fe9a
No special exception type is thrown, after discussion with @per-samuelsson
